src/smt/theory_str_noodler/decision_procedure.cpp

```diff
@@ -617,112 +617,321 @@ namespace smt::noodler {
+     *  - 'c >= 0' should force that c is equal to the s_i!to_code which has '|s_i| == 1', because all others should be -1
+     * 
+     * s = from_code(c)
+     *  - we have solutions for the result s, from this we can create LIA formula restricting the possible values of c
+     *  - we can do the same thing as for to_code, but the first conjunct (|s| != 1 && c == -1) in (1) is replaced with
+     *      (|s| == 0 && c is not a valid code point)
+     *  - we assume that s is restricted to only values that can come from "from_code" (empty string or some char)
+     *      - should be done during processing of from_code in theory_str_noodler, by adding a regular constraint
+     *      - basically means that |s| <= 1
+     * 
+     * i = to_int(s)
+     *  - same as to_code, we want to encode into LIA the possible values of i
+     *  - again, s can be substituted: s = s_1 ... s_n, each s_i can be shared with variables from different to_int (or even to_code/from_code)
+     *  - for each s_i, we create an int variable s_i!to_int, encoding the possible values of s_i as int
+     *  - take each word w = w_1 ... w_n, where w_i is the word of the language L_i of the automaton for s_i (we assume finite L_i, otherwise ERROR)
+     *      - create conjunction C of following conjuncts
+     *              |s_i| == |w_i| && s_i!to_int = to_int('1'.w_i) [ && s_i!to_code = to_code(w_i) ]
+     *          - last part in [] is optional, happens only if s_i substitutes some s used in to_code/from_code
+     *          - to_int('1'.w_i) and to_code(w_i) can be computed, as w_i is a literal, not a variable
+     *          - we add 1 at the beginning of s_i!to_int, because w_i can potentionally start with 0, but we want to encode the exact value of w_i ("005" and "5" should be taken as two different words)
+     *          - if w_i is empty/contains non-digits, then to_int('1'.w_i) returns -1 (we do not need to differentiate between non-valid inputs)
+     *      - add to C
+     *              |s| == |w| && i == to_int(w)                                                                                            (2)
+     *          - again, to_int(w) returns -1 for non-valid input (w is empt/contains non-digits)
+     *      - the cases where w/w_i 
+     *  - final LIA formula is taken as a disjunction of all the conjunctions from the previous step
+     * 
+     * s = from_int(i)
+     *  - similarly to from_code, we want to restrict the values of the argument i from the possible valuations of result s
+     *  - we do the same thing as to_int, but instead of 'i == to_int(w)' in (2) for non-valid w, we put 'i < 0'
+     *  - we assume (as in from_code) that s is restricted to only possible outputs of from_int, mainly that w cannot start with 0 and the only non-valid w is empty string
+     */
     LenNode DecisionProcedure::get_formula_for_conversions() {
```

Comment:
This method is hell. Can you split it into several methods to be more digestible?

Reply:
Agreed. This method really needs to be split into *numerous* smaller methods and functions.

Reply:
It should be split now.